### DISCUSSION TITLE
Metadata Updates

### DISCUSSION THREAD

**Discussion by jaydesl** [DIGITbrain/digitbrain.github.io#3] [Announcements]
# Updates to the Metadata Spec

We'll try to keep everyone up to date with changes via this thread. Feel free to start a discussion around any one change by replying to it below.

Jay

**Comment by jaydesl** [comment]
Let's start with an easy one.

## Support Contextualisation of Cloud Instances

**Asset affected:** Deployments
**Kind of change:** Addition
**Link to modification:** https://github.com/DIGITbrain/digitbrain.github.io/commit/0e3ba72565cc1f67e4692dcd127a897a0fe4e461#diff-efd1cfa654a6e71a89d782c30fafd8665887db9d7fcdf1442c586c3188d6888f

This was a fairly necessary change to add "context" to VMs we create in CloudBroker. With contextualisation we can create files, install libraries or execute ad-hoc shell commands, all well before the application is ever deployed to the node.

**Comment by jaydesl** [comment]
## Move DataSourceMapping to DMA Tuple

**Assets affected:** _Deployments_, _DMA Tuple_
**Kind of change:** Move & removal
**Link to modification:** https://github.com/DIGITbrain/digitbrain.github.io/commit/74d68dc62bd8791d5dd62f7e52c8d68646c2dc29

### About DataSourceMapping (Background)

The original idea behind **DataSourceMapping** was that each _Microservice_ in an _Algorithm_ (which would eventually be part of a _DMA Tuple_) would require specific _Data_ assets (that would also be part of that same _DMA Tuple_). In the original Metadata Spec, this field was placed under _Deployments_.

### Motivations for Change

There are two stages to this change, and two respective motivations.

* The first stage was to move **DataSourceMapping** out of _Deployments_. One _DMA Tuple_ could have multiple _Deployments_, but there would never be a need to define multiple **DataSourceMapping**s for a single _DMA Tuple_. Having **DataSourceMapping** under _Deployments_ therefore made no sense.
* The second stage came from the realisation that we had a **D Assets** field already. We couldn't imagine a case where a link between _Microservice_ and _Data_ asset would **not** exist, so why not turn the **D Assets** field (previously a list) into a mapping of _Microservice_ to _Data_ asset (the role previously covered by **DataSourceMapping**)?

**Comment by jaydesl** [comment]
## Separate fields for Model repository, path and filename

**Asset affected:** Model
**Kind of change:** Change
**Link to modification:** https://github.com/DIGITbrain/digitbrain.github.io/commit/6efffa240feae8618eb5915894b1cabe35a31550#diff-83d9a23133860a3f5583602d397a807f6652d3894eb307c50771e63da11907d8

Proposes a finer grain for specifying the location of a Model file. The location of a Model file **must** now be specified by three mandatory fields:
- `REPOSITORY_URI` - The URI of the repository where the Model file is stored, not including path or filename.
- `PATH` - The path to where the Model file is stored in the specified repository, not including filename.
- `FILENAME` - The filename of the Model at the repository and path above (including file extension if it exists).

**Reply by milanJovicSupsi** [reply]
Dear all, we fully agree with the above modification and support going into this direciton. It's actually in line with the comment of Davide last time. Having the PATH separated from the REPOSITORY_URI gives the important flexibility to keep model path a part.

**Reply by ahajnal** [reply]
Just to avoid confusion, a URI is more general means to refer to a "thing". In Data metadata (and in general), DATA_URI is of the form: SCHEME://HOST:PORT/PATH/FILENAME, where PORT and FILENAME and even PATH is optional in case of Data resources. 
You can split a URI into HOST, PATH, FILENAME, or you can have separate fields, but avoid using REPOSITORY_URI notion if it is only a hostname/IP of a server.

**Comment by jaydesl** [comment]
## Make DataSourceMapping a Substructure

**Assets affected:** _DMA Tuple_
**Kind of change:** Move and modify
**Link to modification:** https://github.com/DIGITbrain/digitbrain.github.io/commit/a415ff1562c68dabc451bb65f22bdbbc37f50b99
**Related:** https://github.com/DIGITbrain/digitbrain.github.io/discussions/3?sort=new#discussioncomment-2107084

Having moved DataSourceMapping to DMA Tuple, it was decided that it should be extracted into a separate _substructure_ for the benefit of users and developers. **DataSourceMapping** is still a key under DMA Tuple, but it is now [documented separately](https://digitbrain.github.io/dataassetsmapping/) for more clarity.

Additionally, the structure of the map has been modified. For each Microservice, Data Sources specified therein (see [Input Data](https://digitbrain.github.io/attributes/microservice/#inputdata) and
[Output Data](https://digitbrain.github.io/attributes/microservice/#outputdata)) must be mapped to a Data Asset.

See the following example:

```json
"DataAssetsMapping": {
    "microservice_aaa_uuid": {
        "my_kafka_out": "data_zzz_uuid",
        "my_file_out": "data_yyy_uuid"
    }
    "microservice_bbb_uuid": {
        "my_stream_in": "data_zzz_uuid"
    }
}
```

**Reply by ahajnal** [reply]
In the example, as far as I understood, instead  of "data_zzz" there should be the UUID of a data resource.
"my_kafka_out" is a parameter - more precisely, an "output data" element's "output_id" - in the Microservice's metadata.
In Microservice metadata, there are separate "inputData" and "outputData" sections, which is redundant, given that DATA_DIRECTION implies input/output. Can we contract them into one section, e.g., "data resources"?

**Reply by jaydesl** [reply]
Yes, I believe you have the correct interpretation. I've modified the example slightly to better show which is a UUID and which is not. See the complete example here: https://digitbrain.github.io/dataassetsmapping/#an-example

Regarding the redundancy of inputData / outputData I happily agree to contract these into a single section. Let's discuss it this week.

**Reply by jaydesl** [reply]
Here is a preview of what the spec will look like with inputData and outputData condensed into a single dataResource:
https://github.com/DIGITbrain/digitbrain.github.io/compare/main...cd3b99648e0b54e44665e6b68146d7325bd6dc50#diff-9ef6d19e3ad9a7402b39bf8b78154903d376594a56c0923eeb90727b6613c2fd
